swap/src/bin/asb.rs

```diff
@@ -98,8 +98,13 @@ async fn main() -> Result<()> {
             let kraken_price_updates = kraken::connect()?;
 
             let mut swarm = swarm::alice(&seed)?;
-            Swarm::listen_on(&mut swarm, config.network.listen)
-                .context("Failed to listen network interface")?;
+
+            for listen in config.network.listen {
+                Swarm::listen_on(&mut swarm, listen.clone()).context(format!(
+                    "Failed to listen on network interface {}",
+                    listen.clone()
+                ))?;
```

Comment:
Minor nit, but when using `format!`, you should always use `with_context` to avoid the allocation that is happening as part of `format!` in case there is no error.

Why?

Before you call a function, all its arguments need to be evaluated. Hence, with `.context` and `format!`, the `format!` needs to evaluated before `.context` can even be executed.
However, if there is no error (i.e. `listen_on` succeeds), then the argument passed to `.context` will never be needed because the internal control flow doesn't need this error message if the whole thing succeeds.

Using `with_context` avoids this because all that needs to be evaluated is the closure that is being passed to `with_context`. This closure acts as a layer of indirection that will only be called when needed (read: if there is an error) and hence also the allocation within `format!` only happens when really necessary.

Additionally, I am surprised you need to `.clone()` it again within the `format!` macro.